File: mockspresso-api/src/main/java/com/episode6/hackit/mockspresso/api/InjectionConfig.java
package com.episode6.hackit.mockspresso.api;

import com.episode6.hackit.mockspresso.reflect.TypeToken;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.annotation.Annotation;
import java.lang.reflect.Constructor;
import java.util.List;

/**
 * A configuration interface that defines how we construct real objects.
 */
public interface InjectionConfig {

  interface ConstructorSelector {
    @Nullable <T> Constructor<T> chooseConstructor(@NotNull TypeToken<T> typeToken);
  }

  @NotNull ConstructorSelector provideConstructorSelector();
  @NotNull List<Class<? extends Annotation>> provideInjectableFieldAnnotations();
  @NotNull List<Class<? extends Annotation>> provideInjectableMethodAnnotations();
}

File: mockspresso-api/src/main/java/com/episode6/hackit/mockspresso/api/InjectionConfig.kt
package com.episode6.hackit.mockspresso.api

import com.episode6.hackit.mockspresso.reflect.TypeToken
import java.lang.reflect.Constructor

/**
 * A configuration interface that defines how we construct real objects.
 */
interface InjectionConfig {
  fun provideInjectableFieldAnnotations(): List<Class<out Annotation>>
  fun provideInjectableMethodAnnotations(): List<Class<out Annotation>>
  fun <T> chooseConstructor(typeToken: TypeToken<T>): Constructor<T>?
}
